### PR TITLE
Add options for dumping PIR/PLC

### LINKED ISSUE #110: Add 'Dump Scripts' test option via test provider/ingredient
For various workflows, It would be pretty useful to have an option to dump PlutusIR, Plutus Core, and Untyped Plutus Core of scripts involved in `tasty-plutus` tests, including the size tests.

I have a rough mockup which dumps uplc to files when the `--dump-scripts path` option is set, using a test-provider for IO, requiring manual annotation of tests via `withDumpScripts`.

My mockup ([Gist](https://gist.github.com/t1lde/f1becd42c1236a32f608f7e6e289bd3b))

Some additions/changes that could be made:
- It would be nicer to have this built into `WithScript` directly.
- Better options UX
- Dumping PIR and typed PLC - this would require the `CompiledCode` values rather than the `Script` values I'm using here (WithScript would also be convenient here)
- It might be better as a Tasty Ingredient, rather than a test Provider

### PR DESCRIPTION
**Pull request by re-xyr**
WIP. Fixes #110. Approach may change, docs haven't been updated yet.

### CODE REVIEW

**Comment by re-xyr** [comment]
`tasty` is a bit inflexible in this situation. We cannot use test ingredients, as they cannot inspect the types of tests. We *can* use a separate test a la the original prototype in #110, but that is in principle not a test. But this is the only working method so far, so I'll use that.

**Comment by kozross** [comment]
This seems fine in principle. @t1lde - is this solution OK for you?

**Comment by L-as** [comment]
It doesn't seem like you're prettifying the output. You could take inspiration from https://github.com/Plutonomicon/plutarch/blob/24d4f63484169ff65edca6cb01fddb3222bec09b/Plutarch.hs#L55

**Comment by re-xyr** [comment]
@L-as Thanks, I've added them to the PLC dumping. PIR does not seem to have a `Readable` prettyprinter though so I used the `Classic` one.

File: tasty-plutus/src/Test/Tasty/Plutus/Internal/DumpScript.hs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  forall (v :: Type).
+  (IsOption v, Coercible v DumpPath) =>
+  OptionSet ->
+  (FilePath -> IO ()) ->
+  IO ()
+runIf opts action = case coerce (lookupOption @v opts) of
+  DumpPath path -> action path
+  NoDumpPath -> pure ()
+
+data DumpScript = forall a. DumpScript String Script (CompiledCode a)
```
Comment (kozross) [inline tasty-plutus/src/Test/Tasty/Plutus/Internal/DumpScript.hs:40]: I'd prefer this written as a GADT:

```haskell
data DumpScript where
      DumpScript :: String -> Script -> CompiledCode a -> DumpScript
```

**Review by kozross — CHANGES_REQUESTED** [review]
This seems good. The README should probably be updated so that someone who just found this package would know that we offer this functionality.

File: tasty-plutus/src/Test/Tasty/Plutus/Internal/DumpScript.hs (hunk 1 above)
Comment (L-as) [reply tasty-plutus/src/Test/Tasty/Plutus/Internal/DumpScript.hs:40]: Out of curiosity: Why?

Comment (kozross) [reply tasty-plutus/src/Test/Tasty/Plutus/Internal/DumpScript.hs:40]: Mostly because I find it easier. I'm not particularly opinionated on this though.

**Review by t1lde — APPROVED** [review]
:+1: LGTM, this will come in handy

**Review by kozross — APPROVED** [review]

**Comment by kozross** [comment]
@re-xyr - please fix the conflicts so we can merge this.

**Comment by re-xyr** [comment]
Thanks to v6.0 API change, this dumping functionality doesn't need to change the API anymore; it should be a minor version.

**Pull request merged**